src/GNDStk/gnds-tree.hpp

```diff
@@ -0,0 +1,644 @@
+            form, "'<'", "XML");
+   } else if (is.peek() == 137) {
+      // looks like hdf5
+      if (form == format::null)
+         form = format::hdf5;
+      else if (form != format::hdf5)
+         detail::warning_tree_io_data<char>(
+            form, "char 137", "HDF5");
+   } else {
+      // looks like json (via process of elimination)
+      if (form == format::null)
+         form = format::json;
+      else if (form != format::json)
+         detail::warning_tree_io_data<char>(
+            form, "neither '<' nor char 137", "Json");
+   }
+
+   // The above logic is such that form cannot henceforth be format::tree,
+   // which would have triggered a return, or format::null, which would have
+   // become one of {xml,json,hdf5} somewhere in the above conditional.
+
+   // ------------------------
+   // Read and convert
+   // ------------------------
+
+   // Obey form, independent of whatever might or might not have been warned
+   // about above. Note that if the original form parameter was format::null,
+   // then form would have been modified, above, to the likely correct format,
+   // based on the peek() calls.
+   if (form == format::xml) {
+      // assume .xml
+      // go through a temporary xml object to create the tree...
+      const xml tmp(is);
+      if (!is.fail())
+         convert(tmp, *this);
+   } else if (form == format::json) {
+      // assume .json
+      // go through a temporary json object to create the tree...
+      const json tmp(is);
+
+      // It would seem that the nlohmann::json stream input operation,
+      // which is used by the constructor we just called, sets failbit
+      // in instances in which it should just set eofbit. So, for now,
+      // we'll comment-out the !is.fail() test... :-/
+      // if (!is.fail())
+      convert(tmp, *this);
+   } else if (form == format::hdf5) {
+      error("HDF5 read() is not implemented yet");
+   } else {
+      // we may or may not want slippery-slope stuff like this
+      internal("Unrecognized form = " + std::to_string(int(form)) + " "
+               "in tree.read()");
+   }
+
+   // done
+   return is;
+}
+
+
+
+// operator>>
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+inline std::istream &operator>>(
+   std::istream &is,
+   Tree<MCON,CCON> &obj
+) {
+   // Call read(istream) above
+   return obj.read(is);
+}
+
+
+
+// -----------------------------------------------------------------------------
+// write
+// -----------------------------------------------------------------------------
+
+// write(char *, format)
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool Tree<MCON,CCON>::write(
+   const char * const file,
+   const format _form
+) const {
+
+   // ------------------------
+   // format::null
+   // Decide from file name
+   // ------------------------
+
+   format form = _form;
+   if (form == format::null) {
```

Comment:
I need to go look carefully at the specific place you're referring to, when I have more time to look. Offhand, I'm sure it's either because I judged that splitting up some code that *could* be interleaved gives better clarity in this case (that sometimes happens), or there's another good reason I split the two. It may relate to an issue with file names versus streams; and with file magic numbers versus file extensions. I'll take a closer look when I get a chance.

Reply:
I believe everything here is correct. If something is still unclear, let me know.